prog/setupgeom/parser_setup.F90

```diff
@@ -0,0 +1,641 @@
+    call getChildValue(root, "Task", pTask, child=pTaskType, default='uploadcontacts')
+    call getNodeName(pTask, buffer)
+
+    if (char(buffer).ne."setupgeometry") then
+      call getChild(root, "Device", pDevice)
+      call getChildValue(pDevice, "AtomRange", transpar%idxdevice)
+    end if
+    
+    call getChildren(root, "Contact", pNodeList)
+    transpar%ncont = getLength(pNodeList)
+    if (transpar%ncont < 2) then
+      call detailedError(root, "At least two contacts must be defined")
+    end if
+    allocate(transpar%contacts(transpar%ncont))
+
+    select case (char(buffer))
+    case ("setupgeometry")
+      
+      call readContacts(pNodeList, transpar%contacts, geom, char(buffer), iAtInRegion, nPLs)
+      call getSKcutoff(pTask, geom, skCutoff)
+      write(stdOut,*) 'Maximum SK cutoff:', SKcutoff
+      call getChildValue(pTask, "printInfo", printDebug, .false.)
+      call setupGeometry(geom, iAtInRegion, transpar%contacts, skCutoff, nPLs, printDebug)
+
+    case default
+
+      call getNodeHSDName(pTask, buffer)
+      call detailedError(pTaskType, "Invalid task '" // char(buffer) // "'")
+
+   end select
+
+   call destroyNodeList(pNodeList)
+
+  end subroutine readTransportGeometry
+
+
+  !> Read bias information, used in Analysis and Green's function eigensolver
+  subroutine readContacts(pNodeList, contacts, geom, task, iAtInRegion, nPLs)
+    type(fnodeList), pointer :: pNodeList
+    type(ContactInfo), allocatable, dimension(:), intent(inout) :: contacts
+    type(TGeometry), intent(in) :: geom
+    character(*), intent(in) :: task
+    type(WrappedInt1), allocatable, intent(out) :: iAtInRegion(:)
+    integer, intent(out), allocatable :: nPLs(:)
+
+    real(dp) :: contactLayerTol, vec(3)
+    integer :: ii, jj
+    type(fnode), pointer :: field, pNode, pTmp, pWide
+    type(string) :: buffer, modif
+    type(listReal) :: fermiBuffer, vecBuffer
+    integer, allocatable :: tmpI1(:)
+
+    allocate(iAtInRegion(size(contacts)+1))
+    allocate(nPLs(size(contacts)))
+
+    do ii = 1, size(contacts)
+
+      contacts(ii)%wideBand = .false.
+      contacts(ii)%wideBandDos = 0.0_dp
+
+      call getItem1(pNodeList, ii, pNode)
+      call getChildValue(pNode, "Id", buffer, child=pTmp)
+      buffer = tolower(trim(unquote(char(buffer))))
+      if (len(buffer) > mc) then
+        call detailedError(pTmp, "Contact id may not be longer than " // i2c(mc) // " characters.")
+      end if
+      contacts(ii)%name = char(buffer)
+      if (any(contacts(1:ii-1)%name == contacts(ii)%name)) then
+        call detailedError(pTmp, "Contact id '" // trim(contacts(ii)%name) &
+            &//  "' already in use")
+      end if
+
+      call getChildValue(pNode, "PLShiftTolerance", contactLayerTol, 1e-5_dp, modifier=modif,&
+          & child=field)
+      call convertByMul(char(modif), lengthUnits, field, contactLayerTol)
+
+      if (task .eq. "setupgeometry") then
+        call getChildValue(pNode, "NumPLsDefined", nPLs(ii), 2)    
```

Comment:
Remove the ",2" default, so that this line becomes obligatory in the input file (instead of assuming 2 PLs, require the user to think about this).